Utils and libraries: Removes the boost/algorithm/string/join dependency

l2a5b1 opened the pull request:
This commit removes the `boost/algorithm/string/join` dependency from the project by replacing `boost::algorithm::join` with the helper function proposed by @MarcoFalke in https://github.com/bitcoin/bitcoin/pull/13726#discussion_r204159967

Empact commented:
Given it's only used in the one place, I'd be inclined to make it static to the file used, or making a more narrow util file for it.

MarcoFalke commented on src/utilstrencodings.h:
```diff
@@ -173,4 +174,24 @@ bool ConvertBits(const O& outfn, I it, I end) {
+    }
+    return stream.str();
+}
+
```
Instead of a whole util function you might just inline a helper to append the warning:

```cpp
static void AppendWarn(std::string& res, const std::string& warn){
if (!res.empty()) res+=", ";
res+=warn;
}

l2a5b1 committed:
Removes the boost/algorithm/string/join dependency

This commit removes the `boost/algorithm/string/join` dependency
from the project by replacing `boost::algorithm::join` with
a simple helper function.

251Labs commented:
Thanks @MarcoFalke and @Empact, appreciate the feedback! 

I have implemented the proposed helper function in f94ec7c and updated the pull request description accordingly.

Empact commented on src/validation.cpp:
```diff
@@ -2293,7 +2299,7 @@ void static UpdateTip(const CBlockIndex *pindexNew, const CChainParams& chainPar
       GuessVerificationProgress(chainParams.TxData(), pindexNew), pcoinsTip->DynamicMemoryUsage() * (1.0 / (1<<20)), pcoinsTip->GetCacheSize());
     if (!warningMessages.empty())
-        LogPrintf(" warning='%s'", boost::algorithm::join(warningMessages, ", ")); /* Continued */
+        LogPrintf(" warning='%s'", warningMessages.c_str()); /* Continued */
```
I don't think `c_str` is needed here.

Empact commented:
utACK 5f019d5

251Labs commented:
Thanks @Empact, yeah your right. I didn't expect that one 👍 
Addressed in 5f019d5.

practicalswift commented:
utACK 5f019d5354cb12e343ea4bb88da04fbe0e98f102

Thanks for helping remove our Boost dependencies! Please continue :-)

sipa commented:
utACK 5f019d5354cb12e343ea4bb88da04fbe0e98f102

MarcoFalke commented:
utACK 5f019d5
